### PR TITLE
[feat] Implementation for OP_OVER & OP_2OVER opcodes 

### PR DESCRIPTION
**Pull request by mubarak23**
- [x] issue #69 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/shinigami/blob/main/CONTRIBUTING.md)
- [x] code change includes tests

Implement OP_OVER & OP_2OVER opcodes

### CODE REVIEW

**Comment by mubarak23** [comment]
the test are failing, @TAdev0 and @b-j-roberts what do you think could be wrong with the test, 

am getting the following error on both of the two test


```
shinigami::opcodes::tests::test_stack::test_opcode_over - Panicked with "Execution of the program failed".
   shinigami::opcodes::tests::test_stack::test_opcode_2over - Panicked with "Execution of the program failed".

```

File: src/opcodes/tests/test_stack.cairo [hunk 1]
```diff
@@ -290,3 +290,22 @@ fn test_op_2rot_insufficient_items() {
     let mut engine = utils::test_compile_and_run_err(program, Error::STACK_OUT_OF_RANGE);
     utils::check_dstack_size(ref engine, 5);
 }
+
+#[test]
+fn test_opcode_over() {
+    let program = "OP_OVER";
```
Comment (b-j-roberts) [inline src/opcodes/tests/test_stack.cairo:296]: You'll need to push values onto the stack for op_over to use. I'd recommend using the examples on this page https://en.bitcoin.it/wiki/Script

**Review by b-j-roberts — CHANGES_REQUESTED** [review]

File: src/opcodes/tests/test_stack.cairo (hunk 1 above)
Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:296]: > You'll need to push values onto the stack for op_over to use. I'd recommend using the examples on this page https://en.bitcoin.it/wiki/Script

the script link specify we pass two value (i.e x1 x2) and expect  x1 x2 x1 as output,

which is what i did here 

```
#[test]
fn test_opcode_over() {
    let program = "OP_1 OP_2 OP_OVER";
    let mut engine = utils::test_compile_and_run(program);
    utils::check_dstack_size(ref engine, 3);
    let expected_dstack = array![ScriptNum::wrap(1), ScriptNum::wrap(2), ScriptNum::wrap(1)];
    utils::check_expected_dstack(ref engine, expected_dstack.span());
}
```

i still got the following error,


`
shinigami::opcodes::tests::test_stack::test_opcode_over - Panicked with "Execution of the program failed".
`

could the error came as a result of opcode_over function failing to run successfully

can you inspect the function and let me know if this

Comment (b-j-roberts) [reply src/opcodes/tests/test_stack.cairo:296]: I'd recommend using println! statements to debug. Your test is now setup correct.
Debugging could be good practice, let me know.

Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:296]: Alright

Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:296]: println!("{:?}", bytecode);
    let mut engine = EngineTraitImpl::new(bytecode);
    let res = engine.execute();
    println!("{:?}", res);
    
    return the following result 
    
```
    
"QR\0"
Result::Err(34304024605757925039519833781264315913130327289331821709226243173)
```

this is the opcode program

"OP_1 OP_2 OP_OVER"

Comment (b-j-roberts) [reply src/opcodes/tests/test_stack.cairo:296]: Could you use println! within the over_n implementations to see where the issue is occurring

Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:296]: the code does not reach that point,

both opcode_over and over_n.

Comment (b-j-roberts) [reply src/opcodes/tests/test_stack.cairo:296]: I think you never added the opcodes to the compiler

Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:296]: hmm

you are right

Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:296]: i suspect the while loop am using

Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:296]: test from function [test_opcode_over] is running successfully now,

Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:296]: test_opcode_2over  is also running successfully

**Comment by mubarak23** [comment]
@b-j-roberts 

this test is failing now 
test_op_nop_x_all
with the following result 

```

failures:
   shinigami::opcodes::tests::test_flow::test_op_nop_x_all - Panicked with "assertion `err == expected_err` failed: Program did not return the expected error
err: 34304024605757925039519833781264315913130327289331821709226243173
expected_err: 523853083712021213892934608376371690915972896855266453451877".
```

**Comment by mubarak23** [comment]
Does addition of test function has anything to do with it @b-j-roberts

**Review by b-j-roberts — APPROVED** [review]
Thank you!

**Comment by mubarak23** [comment]
> Thank you!

Welcome 
Looking forward to more contributions

**Pull request merged**